hardware_pydantic/devices.py

```diff
@@ -56,26 +88,97 @@ def action__transfer(
             self,
             from_obj: Rack | Heater,
             to_obj: Rack | Heater,
-            transferee: Vial,
+            vial: Vial,
             to_position: str | None
     ) -> None:
         # TODO make more general
         # take it out
-        assert transferee.position_relative == from_obj.identifier
+        assert vial.position_relative == from_obj.identifier
         if isinstance(from_obj, Rack):
-            assert transferee.position in from_obj.content
-            from_obj.content[transferee.position] = None
+            assert vial.position in from_obj.content
+            from_obj.content[vial.position] = None
         elif isinstance(from_obj, Heater):
             from_obj.content = None
         else:
             raise TypeError
-        transferee.position = None
-        transferee.position_relative = None
+        vial.position = None
+        vial.position_relative = None
         # put it to
         if isinstance(to_obj, Rack):
             assert to_position in to_obj.content
-            to_obj.content[to_position] = transferee.identifier
+            to_obj.content[to_position] = vial.identifier
         elif isinstance(to_obj, Heater):
-            to_obj.content = transferee.identifier
-        transferee.position = to_position
-        transferee.position_relative = to_obj.identifier
+            to_obj.content = vial.identifier
+        vial.position = to_position
+        vial.position_relative = to_obj.identifier
+
+
+class PlateGripperArm(Device):
+    pass
+
+
+class SolidSVTool(Device):
+    """Solid sample vial tool for transferring solids."""
+    # todo: this device is often used together with a balance
+    @action_method_logging
+    def action__transfer(self, from_obj: Vial, to_obj: Vial, amount: float):
+        # TODO sample from a dist
+        removed = from_obj.remove_content(amount)
+        to_obj.add_content(removed)
+
+    def projection__transfer(self, from_obj: Vial, to_obj: Vial, amount: float):
+        # todo: add adaptive projection to estimate the time of transferring solids
+        transfer_speed = 2
+        return amount / transfer_speed
+
+
+class Balance(Device):
+    """Balance for weighing solids."""
+    reading: float = 0
+    content: LabObject | None = None
+
+    @action_method_logging
+    def action__tare(self):
+        self.reading = 0
+
+    def projection__tare(self):
+        return 0
+
+    @action_method_logging
+    def action__weigh(self, amount: float):
+        self.reading = amount
+
+    def projection__weigh(self, amount: float):
```

Comment:
What is the `projection__weigh` action and why is there an `amount` for a balance?

Reply:
This class method tells us how long it takes to weigh the sample. 

My original thought: the more sample we have, the longer the balance will have to wait. For example, weighing 10 mg and 50 g will take different time. But now I realized that this timing should not belong to the balance, but the liquid or solid transfer device. Therefore, I took out the `amount` and treated the timing of weighing as a constant. 

Fixed in 0d19b0f.